cmd/controller/app/controller.go

```diff
@@ -97,7 +98,7 @@ func Run(opts *options.ControllerOptions, stopCh <-chan struct{}) error {
 	})
 
 	// Start profiler if it is enabled
-	if opts.EnablePprof {
+	if *opts.EnablePprof {
```

Comment:
How do we know that this is a non-nil pointer?

Reply:
[NewControllerConfiguration](https://github.com/AcidLeroy/cert-manager/blob/controller-configuration-file/cmd/controller/app/options/options.go#L58-L70) will create a ControllerConfiguration object with [sane defaults](https://github.com/AcidLeroy/cert-manager/blob/controller-configuration-file/internal/apis/config/controller/v1alpha1/defaults.go#L177-L324), ensuring that values are not nil. Furthermore, this is verified in the unit tests when running the [fuzzer](https://github.com/AcidLeroy/cert-manager/blob/controller-configuration-file/internal/apis/config/controller/install/roundtrip_test.go)

Reply:
Would it be good-practice to make the internal types non-optional & do defaulting for the public types only (before converting to the internal types)?

Reply:
That's a good question, I'm not entirely sure how to answer that? So you are saying that we should  *not* include any pointers for the internal type? I.e., everything in the [config/controller/types.go](https://github.com/cert-manager/cert-manager/blob/bd2fdd728c3015d8f45727408242fcb75cbe53f8/internal/apis/config/controller/types.go) file should *not* have a pointer? Looking at the Kubelet code, they do have a few pointers in their internal types ([MaxImagePulls](https://github.com/kubernetes/kubernetes/blob/f61a0e41075c2c4468e6cda84f7a88a960fb2933/pkg/kubelet/apis/config/types.go#L294-L295) for example). 

Maybe you can point me to an example of what you are thinking of?  At the end of the day, you are worried about dereferencing a null pointer in this file?

Reply:
Ok, I think I understand what you are saying. I think I kept the internal and the public facing types the same. I only did it that way because I thought that was the pattern. But I see now that you've pointed out the fact that the public API may have optional fields (i.e. pointers) but they are eventually defaulted and translated to non-pointer fields in the internal types. I'll have to look at the original controller code and see if any fields _should_ be optional. I'll update the PR and ping you when I get it finished.

Reply:
Thank you very much @AcidLeroy and sorry for giving your more work 😁 .

Reply:
@inteon so I had a little bit of time to look into this today, and what I found was that the conversion-gen function has no conversion, for a few of the types that are currently in the controller. For example, I've changed the internal type of `KubernetesAPIQPS` from a `*float32` to a `float32`. When I ran the conversion too on this, I found that there were some warnings: 

```
// WARNING: in.KubernetesAPIQPS requires manual conversion: inconvertible types (*float32 vs float32)
```

I thought that was a bit strange! So I investigated, and sure enough there is no conversion for that type. There is however, a conversion from a [*float64 to a float64](https://github.com/kubernetes/apimachinery/blob/master/pkg/apis/meta/v1/conversion.go#L33). Does that seem strange to you? At any rate, the conversion-gen function can't handle that type automatically. So I am left with the following options: 
1. Write a manual conversion for that type, preserving the type, or 
2. Change the types that are not supported into ones that are. 

I'm leaning towards 2, just because it won't change the existing API. However, it seems like there will be more conversion to manage as a result of this. 

What are your thoughts on that?

Reply:
> All public integer fields MUST use the Go int32 or Go int64 types, not int (which is ambiguously sized, depending on target platform). Internal types may use int.

> For integer fields, prefer int32 to int64 unless you need to represent values larger than int32. See other guidelines about limitations of int64 and language compatibility.

source: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md

It seems like `*int32` would be best for the external types, I'm unsure what happens if you change the internal type to `int` instead of `int32`.

Reply:
I don't see float32 being mentioned anywhere in the document, I guess using float64 won't do any harm?

Reply:
@AcidLeroy I think we should do all conversions in the conversion functions, so the internal types should match the types used in our code and the external types should follow best practices (int32 or int64). If there are any conversion functions missing, we should add them.

Reply:
hmmm, what do you think we should do about the floats, given that they are not recommended? Should I just add some more bits and go directly to `float64`? I'll do the int to int32 conversion and see if there are any round-trip errors. 

If there are round tripping problems, what you do you think the best way to proceed is? Should I change the internal types to reflect those recommendations? Alternatively, maybe we can accept some precision loss for the fields and handle that at a later time?

Reply:
QPS could actually be expressed as a Hz quantity (k8s.io/apimachinery/pkg/api/resource Quantity).
But that would maybe be a bit overengineered.

Reply:
@AcidLeroy I think it is your decisions whether you want to pursue a more perfect solution still, we can still improve the solution later.

Reply:
@inteon Ok, I made some custom conversion functions for the types that we discussed. I left the float as a float32 for now, but maybe we should investigate in another MR if we should move to something else? Anyway, take a look at the latest commit and see if that satisfies you. https://github.com/cert-manager/cert-manager/commit/acd1d49022ba8770eeba6ea535ff77d5d77be309

Reply:
Amazing work! I like it a lot!